iroh-net-report/src/lib.rs

```diff
@@ -206,6 +207,140 @@ impl Default for Reports {
     }
 }
 
+/// Options for running probes
+///
+/// By default, will run icmp over IPv4, icmp over IPv6, and Https probes.
+///
+/// Use [`Options::stun_v4`], [`Options::stun_v6`], and [`Options::quic_config`]
+/// to enable STUN over IPv4, STUN over IPv6, and QUIC address discovery.
+#[derive(Debug)]
+pub struct Options {
+    /// Socket to send IPv4 STUN probes from.
+    ///
+    /// Responses are never read from this socket, they must be passed in via internal
+    /// messaging since, when used internally in iroh, the socket is also used to receive
+    /// other packets from in the magicsocket (`MagicSock`).
+    ///
+    /// If not provided, STUN probes will not be sent over IPv4.
+    stun_sock_v4: Option<Arc<UdpSocket>>,
+    /// Socket to send IPv6 STUN probes from.
+    ///
+    /// Responses are never read from this socket, they must be passed in via internal
+    /// messaging since, when used internally in iroh, the socket is also used to receive
+    /// other packets from in the magicsocket (`MagicSock`).
+    ///
+    /// If not provided, STUN probes will not be sent over IPv6.
+    stun_sock_v6: Option<Arc<UdpSocket>>,
+    /// The configuration needed to launch QUIC address discovery probes.
+    ///
+    /// If not provided, will not run QUIC address discovery.
+    quic_config: Option<QuicConfig>,
+    /// Enable icmp_v4 probes
+    ///
+    /// On by default
+    icmp_v4: bool,
+    /// Enable icmp_v6 probes
+    ///
+    /// On by default
+    icmp_v6: bool,
+    /// Enable https probes
+    ///
+    /// On by default
+    https: bool,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            stun_sock_v4: None,
+            stun_sock_v6: None,
+            quic_config: None,
+            icmp_v4: true,
+            icmp_v6: true,
+            https: true,
+        }
+    }
+}
+
+impl Options {
+    /// Create an empty Options that enables no probes
+    pub fn empty() -> Self {
+        Self {
+            stun_sock_v4: None,
+            stun_sock_v6: None,
+            quic_config: None,
+            icmp_v4: false,
+            icmp_v6: false,
+            https: false,
+        }
+    }
+
+    /// Set the ipv4 stun socket and enable ipv4 stun probes
+    pub fn stun_v4(mut self, sock: Option<Arc<UdpSocket>>) -> Self {
+        self.stun_sock_v4 = sock;
+        self
+    }
+
+    /// Set the ipv6 stun socket and enable ipv6 stun probes
+    pub fn stun_v6(mut self, sock: Option<Arc<UdpSocket>>) -> Self {
+        self.stun_sock_v6 = sock;
+        self
+    }
+
+    /// Enable quic probes
+    pub fn quic_config(mut self, quic_config: Option<QuicConfig>) -> Self {
+        self.quic_config = quic_config;
+        self
+    }
+
+    /// Enable icmp_v4 probe
+    pub fn enable_icmp_v4(mut self) -> Self {
```

Comment:
It is slightly odd that the `stun_v4`, `stun_v6` and `quic_config` methods all allow both enabling and disabling but then you have this instead of `icmp_v4(mut self, enable: bool)`.

But it doesn't matter much.  This meets our needs anyway.

Reply:
nah ur right it's weird